### PR TITLE
Add ClientFactory to TCP input source to add SplitFunc/NetworkFuncs per client

### PR DESCRIPTION
**Pull request by vjsamuel**
This PR changes the way TCP input source is created to a ConnectionFactory based model which takes a `SplitFunc`, `NetworkFunc` and two `ClientCallback` funcs that can be called during a client connect and disconnect. 

These hooks allow the TCP implementer to do better logging, stateful processing and also spin up splitters per client.

### CODE REVIEW

File: filebeat/inputsource/tcp/client.go [hunk 1]
```diff
@@ -79,24 +95,24 @@ func (c *client) handle() error {
 	for scanner.Scan() {
 		err := scanner.Err()
 		if err != nil {
-			// we are forcing a close on the socket, lets ignore any error that could happen.
+			// we are forcing a Close on the socket, lets ignore any error that could happen.
 			select {
 			case <-c.done:
 				break
 			default:
 			}
```
Comment (urso) [inline filebeat/inputsource/tcp/client.go]: The channel could be closed anytime, not only if the reader fails.
The way this function is written the callback if scan could produce an 'event', but channel is closed in the meantime. Is this expected behavior or do we rather not want to publish the event on close and guarantee the handler can return immediately.

File: filebeat/inputsource/tcp/conn.go [hunk 2]
```diff
@@ -71,7 +71,7 @@ type DeadlineReader struct {
 }
 
 // NewDeadlineReader returns a new DeadlineReader
-func NewDeadlineReader(c net.Conn, timeout time.Duration) *DeadlineReader {
+func NewDeadlineReader(c net.Conn, timeout time.Duration) io.Reader {
```
Comment (urso) [inline filebeat/inputsource/tcp/conn.go]: Why change the return type here? Prefer to return concrete types, but accept interfaces.

**Comment by urso** [comment]
Jenkins, test this.

**Comment by urso** [comment]
This CI fail seems to be related:

```
16:56:14 command [go test -race -cover -coverprofile /tmp/gotestcover-180254964 github.com/elastic/beats/filebeat/inputsource/tcp]: exit status 1
16:56:14 ==================
16:56:15 WARNING: DATA RACE
16:56:15 Read at 0x00c0000a6c00 by goroutine 12:
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*splitHandler).Close()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/client.go:126 +0x56
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*Server).Stop()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server.go:142 +0x202
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.TestReceiveEventsAndMetadata.func1()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server_test.go:201 +0xa04
16:56:15   testing.tRunner()
16:56:15       /var/lib/jenkins/.gvm/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x163
16:56:15 
16:56:15 Previous write at 0x00c0000a6c00 by goroutine 10:
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*splitHandler).Handle()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/client.go:80 +0x6d
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*Server).run.func1()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server.go:120 +0x339
16:56:15 
16:56:15 Goroutine 12 (running) created at:
16:56:15   testing.(*T).Run()
16:56:15       /var/lib/jenkins/.gvm/versions/go1.12.4.linux.amd64/src/testing/testing.go:916 +0x65a
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.TestReceiveEventsAndMetadata()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server_test.go:157 +0x1979
16:56:15   testing.tRunner()
16:56:15       /var/lib/jenkins/.gvm/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x163
16:56:15 
16:56:15 Goroutine 10 (running) created at:
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*Server).run()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server.go:111 +0x220
16:56:15   github.com/elastic/beats/filebeat/inputsource/tcp.(*Server).Start.func1()
16:56:15       /var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat/inputsource/tcp/server.go:85 +0x80
16:56:15 ==================
16:56:15 --- FAIL: TestReceiveEventsAndMetadata (1.72s)
16:56:15     --- FAIL: TestReceiveEventsAndMetadata/MaxReadBufferReachedUserConfigured (0.01s)
16:56:15         testing.go:809: race detected during execution of test
16:56:15     testing.go:809: race detected during execution of test
16:56:15 FAIL
16:56:15 coverage: 81.6% of statements
16:56:15 FAIL	github.com/elastic/beats/filebeat/inputsource/tcp	2.018s
16:56:15 ?   	github.com/elastic/beats/filebeat/scripts/generator/fileset	[no test files]
16:56:15 ?   	github.com/elastic/beats/filebeat/scripts/generator/module	[no test files]
16:56:18 ?   	github.com/elastic/beats/filebeat/scripts/mage	[no test files]
16:56:18 ok  	github.com/elastic/beats/filebeat/registrar	1.038s	coverage: 17.3% of statements
16:56:19 ok  	github.com/elastic/beats/filebeat/scripts/tester	1.020s	coverage: 14.0% of statements
16:56:19 ok  	github.com/elastic/beats/filebeat/processor/add_kubernetes_metadata	1.024s	coverage: 88.5% of statements
16:56:20 ok  	github.com/elastic/beats/filebeat/util	1.018s	coverage: 71.4% of statements
16:56:20 test failed
16:56:20 make[1]: *** [unit-tests] Error 1
16:56:20 make[1]: Leaving directory `/var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-linux/beat/filebeat/label/linux-immutable/src/github.com/elastic/beats/filebeat'
16:56:20 make: *** [testsuite] Error 2
```

**Comment by vjsamuel** [comment]
@urso @ph i am not able to reproduce the data race on local. can you please help rerun Jenkins?

**Comment by vjsamuel** [comment]
never mind i have updated the PR with the fix

File: CHANGELOG.asciidoc [hunk 3]
```diff
@@ -231,6 +231,7 @@ https://github.com/elastic/beats/compare/v7.0.0-beta1...v7.0.0-rc1[Check the HEA
 - Add ISO8601 timestamp support in syslog metricset. {issue}8716[8716] {pull}10736[10736]
 - Add support for loading custom NetFlow and IPFIX field definitions to netflow input. {pull}10945[10945] {pull}11223[11223]
 - Added categorization fields for SSH login events in the system/auth fileset. {pull}11334[11334]
+- Add ClientFactory to TCP input source to add SplitFunc/NetworkFuncs per client. {pull}8543[8543]
```
Comment (urso) [inline CHANGELOG.asciidoc]: The changelog suggest it's an usr visible feature in 7.0.0-rc1. Please move the changelog entry to: CHANGELOG-developer.next.asciidoc

**Comment by vjsamuel** [comment]
thanks @urso

**Pull request merged**